### ISSUE TITLE
Unable to run Habitica locally in Docker

### ISSUE THREAD

**Issue by TaiFeng** [HabitRPG/habitica#11592]
### General Info
  * UUID: 7d0cdda9-2f70-4843-ba31-22f2e0f3c5fb
  * Browser: n/a
  * OS: Ubuntu 18.04.1 LTS (bionic) amd64
  * docker-compose: version 1.25.0, build 0a186604
  * Docker engine: version 19.03.5, build 633a0ea838

### Description
When strictly following the instructions outlined here: https://habitica.fandom.com/wiki/Setting_up_Habitica_Locally_on_Docker

Both the habitica_server and habitica_client docker images crashed upon startup (immediately after running the commmand "docker-compose -f docker-compose.dev.yml up -d"). 

The entire session was recording via screen capture and can be viewed here: https://youtu.be/nFy9tzYeEcc
The description of the video includes links to notable timestamps.

The list of commands sent in via the console and related outputs can be viewed at this GitHub Gist link: https://gist.github.com/TaiFeng/910424424b7303debea62ca3321023aa

#### Console Errors
server_1  | > habitica@4.122.0 start /code
server_1  | > gulp nodemon
server_1  |
server_1  | [08:07:10] Using gulpfile /code/gulpfile.js
server_1  | [08:07:10] Starting 'nodemon'...
server_1  | [08:07:10] Finished 'nodemon' after 176 ms
server_1  | [08:07:10] [nodemon] 1.19.4
server_1  | [08:07:10] [nodemon] to restart at any time, enter `rs`
server_1  | [08:07:10] [nodemon] watching dir(s): *.*
server_1  | [08:07:10] [nodemon] watching extensions: js,mjs,json
server_1  | [08:07:10] [nodemon] starting `node ./website/server/index.js`
server_1  | npm ERR! code ELIFECYCLE
server_1  | npm ERR! errno 1
server_1  | npm ERR! habitica@4.122.0 start: `gulp nodemon`
server_1  | npm ERR! Exit status 1
server_1  | npm ERR!
server_1  | npm ERR! Failed at the habitica@4.122.0 start script.
server_1  | npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
server_1  |
server_1  | npm ERR! A complete log of this run can be found in:
server_1  | npm ERR!     /root/.npm/_logs/2019-11-27T08_07_13_102Z-debug.log

client_1  | > habitica@4.122.0 client:dev /code
client_1  | > cd website/client && npm run serve
client_1  |
client_1  |
client_1  | > habitica-client@1.0.0 serve /code/website/client
client_1  | > vue-cli-service serve
client_1  |
client_1  | sh: 1: vue-cli-service: not found
client_1  | npm ERR! code ELIFECYCLE
client_1  | npm ERR! syscall spawn
client_1  | npm ERR! file sh
client_1  | npm ERR! errno ENOENT
client_1  | npm ERR! habitica-client@1.0.0 serve: `vue-cli-service serve`
client_1  | npm ERR! spawn ENOENT
client_1  | npm ERR!
client_1  | npm ERR! Failed at the habitica-client@1.0.0 serve script.
client_1  | npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
client_1  | npm WARN Local package.json exists, but node_modules missing, did you mean to install?
client_1  |
client_1  | npm ERR! A complete log of this run can be found in:
client_1  | npm ERR!     /root/.npm/_logs/2019-11-27T08_07_11_049Z-debug.log
client_1  | npm ERR! code ELIFECYCLE
client_1  | npm ERR! errno 1
client_1  | npm ERR! habitica@4.122.0 client:dev: `cd website/client && npm run serve`
client_1  | npm ERR! Exit status 1
client_1  | npm ERR!
client_1  | npm ERR! Failed at the habitica@4.122.0 client:dev script.
client_1  | npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
client_1  |
client_1  | npm ERR! A complete log of this run can be found in:
client_1  | npm ERR!     /root/.npm/_logs/2019-11-27T08_07_11_084Z-debug.log

**Comment by SabreCat** [comment]
I'm reviewing a fix for this in https://github.com/HabitRPG/habitica/pull/11518. Try working from the Dockerfile-Dev there for now!

**Comment by SabreCat** [comment]
Fixed in #11518

**Comment by TaiFeng** [comment]
Thanks @SabreCat . I tried the fix, and the client image appeared to stabilize! Unfortunately, it looks like I'm seeing similar errors for the server in the logs. I've attached my console output here: https://gist.github.com/TaiFeng/61ac3eebe0b4dceadaa245e18011bb43

**Comment by paulcalabro** [comment]
I'm experiencing the same issue as well.

**Comment by SabreCat** [comment]
Hmm, OK. @TaiFeng @paulcalabro do you get any better results if you add this line ahead of `RUN npm install` (as in the production Dockerfile)?

>RUN npm set unsafe-perm true

**Comment by yowkah** [comment]
I just wanted to chime in here. I'm havinfg the exact same error as provided in the logs above for the server-side. The client runs well. Adding the specified line to the Dockerfile-dev didn't result in any change.

I did some troubleshooting and tried to find some extensive logs since the stack-trace isn't very long. One of those steps involved eliminating nodemon. After changing the "start" script in package.json to:
```
    "start": "node ./website/server/index.js",
```
The server started and ran succesfully..

Using nodemon however, nodemon starts, tries to launch the index.js, crashes and refers to a debug log, the logs are as follows:

```log
0 info it worked if it ends with ok
1 verbose cli [ '/usr/local/bin/node', '/usr/local/bin/npm', 'start' ]
2 info using npm@6.12.1
3 info using node@v12.13.1
4 verbose run-script [ 'prestart', 'start', 'poststart' ]
5 info lifecycle habitica@4.125.0~prestart: habitica@4.125.0
6 info lifecycle habitica@4.125.0~start: habitica@4.125.0
7 verbose lifecycle habitica@4.125.0~start: unsafe-perm in lifecycle true
8 verbose lifecycle habitica@4.125.0~start: PATH: /usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/node-gyp-bin:/usr/src/habitica/node_modules/.bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
9 verbose lifecycle habitica@4.125.0~start: CWD: /usr/src/habitica
10 silly lifecycle habitica@4.125.0~start: Args: [ '-c', 'gulp nodemon' ]
11 silly lifecycle habitica@4.125.0~start: Returned: code: 1  signal: null
12 info lifecycle habitica@4.125.0~start: Failed to exec start script
13 verbose stack Error: habitica@4.125.0 start: `gulp nodemon`
13 verbose stack Exit status 1
13 verbose stack     at EventEmitter.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/index.js:332:16)
13 verbose stack     at EventEmitter.emit (events.js:210:5)
13 verbose stack     at ChildProcess.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/lib/spawn.js:55:14)
13 verbose stack     at ChildProcess.emit (events.js:210:5)
13 verbose stack     at maybeClose (internal/child_process.js:1021:16)
13 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:283:5)
14 verbose pkgid habitica@4.125.0
15 verbose cwd /usr/src/habitica
16 verbose Linux 5.3.12-1-MANJARO
17 verbose argv "/usr/local/bin/node" "/usr/local/bin/npm" "start"
18 verbose node v12.13.1
19 verbose npm  v6.12.1
20 error code ELIFECYCLE
21 error errno 1
22 error habitica@4.125.0 start: `gulp nodemon`
22 error Exit status 1
23 error Failed at the habitica@4.125.0 start script.
23 error This is probably not a problem with npm. There is likely additional logging output above.
24 verbose exit [ 1, true ]
```
Basically since the log only says the process exits with a status 1, they are not really helpful. I hope this helps

**Comment by yowkah** [comment]
Turns out the error is hidden outside of the logs.
The cause seems to be the limit to the amount of file watchers by inotify.

im on manjaro, so the line below fixed the issue for me. If you are on linux as well, make sure you use your distro's equivelant of this to get nodemon to work.
```
echo fs.inotify.max_user_watches=524288 | sudo tee /etc/sysctl.d/50-max-user-watches.conf && sudo sysctl --system
```
In the meantime. If i got spare time this weekend, I'll see if there is an effective way to reduce the amount of file-watchers without degrading the development experience.

**Comment by paulcalabro** [comment]
@SabreCat Adjusting the number of file watchers as noted above resolved my issue.

**Comment by SunSparc** [comment]
I checked my inotify watches and they were good, so that did not seem to be the issue for me. What did finally work for me was this command:

`docker-compose -f docker-compose.dev.yml up -d --force-recreate --build`

Which seems to indicate that there was something cached in the old docker image that was not being cleared out.

This tip came from @Squall in the [Aspiring Blacksmiths](https://habitica.com/groups/guild/68d4a01e-db97-4786-8ee3-05d612c5af6f) guild chat.

**Comment by SabreCat** [comment]
By the look of it, this is a duplicate of #10930, and folks running into trouble seem to have resolved it..? Let us know if this is still happening and needs attention in the Habitica dev Dockerfile.
